File: Golang/724. Find Pivot Index.go
// The pivot index is the index where the sum of all the numbers strictly to the left of the index is equal to the sum of all the numbers strictly to the index's right.
// TC: O(N), SC: O(1)

func pivotIndex(nums []int) int {
	var leftSum, sum int
	for _, num := range nums {
		sum += num
	}
	for idx, val := range nums {
		if leftSum == sum-leftSum-val {
			return idx
		}
		leftSum += val
	}
	return -1
}

/*
Input: nums = [1,7,3,6,5,6]
Output: 3
Explanation:
The pivot index is 3.
Left sum = nums[0] + nums[1] + nums[2] = 1 + 7 + 3 = 11
Right sum = nums[4] + nums[5] = 5 + 6 = 11
*/

File: Python/724. Find Pivot Index.py
# O(N), O(1)
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        left, right = 0, sum(nums)
        for idx, num in enumerate(nums):
            right -= num
            if left == right:
                return idx
            left += num
        return -1
